### PR TITLE
add task solution

### PR DESCRIPTION
**Pull request by SerhiiVoitiuk**
[DEMO LINK](https://SerhiiVoitiuk.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,349 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, {
+  useCallback, useEffect, useMemo, useState,
+} from 'react';
```
Comment (loralevitska) [inline src/App.tsx:3]: Fix all cases

```suggestion
import React, {
  useCallback, 
  useEffect, 
  useMemo, 
  useState,
} from 'react';
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,349 @@
...
 
-const USER_ID = 0;
+const USER_ID = 11689;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState('');
+  const [filter, setFilter] = useState(Filters.All);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [pageIsLoaded, setPageIsLoaded] = useState(false);
```
Comment (loralevitska) [inline src/App.tsx:26]: Rename all boolean variables according [tips](https://dev.to/michi/tips-on-naming-boolean-variables-cleaner-code-35ig)

```suggestion
  const [isPageLoading, setIsPageLoading] = useState(false);
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,349 @@
...
+    try {
+      setIsUpdatingId(state => (
+        [...state, id]
+      ));
+      await updateTodo(id, { completed: !completed });
+      setTodos(state => {
+        state.map(todo => {
+          if (todo.id === id) {
+            // eslint-disable-next-line no-param-reassign
+            todo.completed = !completed;
```
Comment (loralevitska) [inline src/App.tsx:123]: Don't ignore eslint rule, try to fix them
```suggestion
            return {
              ...todo,
              completed: !completed,
            }
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,349 @@
...
+    setTodos(stateMain => {
+      stateMain.map(async (todo) => {
+        try {
+          if (isStatus && todo.completed) {
+            setIsUpdatingId(state => (
+              [...state, todo.id]
+            ));
+            await updateTodo(todo.id, { completed: false });
+            // eslint-disable-next-line no-param-reassign
+            todo.completed = false;
```
Comment (loralevitska) [inline src/App.tsx:154]: According previous comment fix eslint rule. Fix all cases if present

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,349 @@
...
+      handleErrorSet(ErrorMessage.updateTodo);
+      setIsUpdatingId((state) => (
+        [...state.filter(stateId => stateId !== id)]
+      ));
+    }
+  };
+
+  const completedTodos = useMemo(() => {
+    return todos.filter(todo => todo.completed).length;
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (loralevitska) [inline src/App.tsx:294]: Remove comments and fix rules. Fix all cases if present

File: src/api/todos.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+export const getTodos = (userId: number) => {
+  return client.get<Todo[]>(`/todos?userId=${userId}`);
+};
+
+export const deleteTodo = (todoId: number) => {
+  return client.delete(`/todos/${todoId}`);
+};
+
+export const addTodo = (url: string, data: any) => {
```
Comment (loralevitska) [inline src/api/todos.ts:12]: Fix all cases with warning also 
image

File: src/components/Footer.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      </nav>
+
+      <button
+        type="button"
+        className="todoapp__clear-completed"
+        data-cy="ClearCompletedButton"
+        disabled={!completedTodos}
+        onClick={() => {
+          handleMultipleDelete();
+        }}
```
Comment (loralevitska) [inline src/components/Footer.tsx:72]:
```suggestion
        onClick={handleMultipleDelete}
```

File: src/components/TodoList.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          completed,
+        } = todo;
+
+        return (
+          <div
+            id={`${id}`}
+            key={id}
+            data-cy="Todo"
+            className={classNames('todo', { completed })}
+          >
```
Comment (loralevitska) [inline src/components/TodoList.tsx:46]: Move this block to separate component `TodoItem`

**Review by loralevitska — CHANGES_REQUESTED** [review]
Please provide demo_link for review 
image

File: src/App.tsx [hunk 9]
```diff
@@ -1,24 +1,336 @@
...
+  const closeError = () => {
+    setErrorMessage('');
+  };
+
+  const loadTodos = useCallback(async () => {
+    try {
+      const loadedTodos = await getTodos(USER_ID);
+
+      setTodos(loadedTodos);
+      setIsPageLoading(true);
```
Comment (denys-danyliuk) [inline src/App.tsx:52]: you should do it before you load todos, also you should eventually turn it to false when todos are loaded

File: src/App.tsx [hunk 10]
```diff
@@ -1,24 +1,336 @@
...
+            todo.completed
+          ));
+        default: return todos;
+      }
+    }
+
+    return todos;
+  };
+
+  const handleFormSubmit = async (e?: Event) => {
```
Comment (denys-danyliuk) [inline src/App.tsx:82]: try to specify the event

File: src/App.tsx [hunk 11]
```diff
@@ -1,24 +1,336 @@
...
+    try {
+      const newTodo = {
+        userId: USER_ID,
+        title: query.trim(),
+        completed: false,
+        id: 0,
+      };
+
+      setTempTodo({ ...newTodo, id: 0 });
+      setIsPageLoading(false);
```
Comment (denys-danyliuk) [inline src/App.tsx:99]: use finally to set loading to false

File: src/App.tsx [hunk 12]
```diff
@@ -1,24 +1,336 @@
...
+        return [...state, newTodos as Todo];
+      });
+
+      setTempTodo(null);
+      setIsPageLoading(true);
+      setQuery('');
+    } catch {
+      handleErrorSet(ErrorMessage.addTodo);
+      setTempTodo(null);
+      setIsPageLoading(true);
```
Comment (denys-danyliuk) [inline src/App.tsx:113]: for what reason you have it here?

File: src/App.tsx [hunk 13]
```diff
@@ -1,24 +1,336 @@
...
+            setIsUpdatingId([]);
+          }
+        });
+
+        return stateMain;
+      });
+    }
+  };
+
+  const ecsKeyCancel = (e: string) => {
```
Comment (denys-danyliuk) [inline src/App.tsx:223]: method name should start with a verb

File: src/components/Footer.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+export const Footer: React.FC<Props> = React.memo(({
+  filter,
+  filterValue,
+  completedTodos,
+  uncompletedTodos,
+  handleDelete,
+}) => {
+  const handleMultipleDelete = () => {
+    handleDelete();
+  };
```
Comment (denys-danyliuk) [inline src/components/Footer.tsx:22]: if this function just calls another function than why do you need this wrapper?

**Review by denys-danyliuk — CHANGES_REQUESTED** [review]
you should revise your logic with loading, variable should have value of true when you are loading or doing something which involves communication with api and after you finish it should become false

**Pull request closed without merging**